hashing/src/chunk.rs

```diff
@@ -85,4 +133,44 @@ mod test {
         assert_eq!(proof_0.root_hash(), expected_root);
         assert_eq!(proof_1.root_hash(), expected_root);
     }
+
+    #[test]
+    fn validates_chunk_with_proofs() {
+        let large_data_size = (CHUNK_SIZE * 2) as usize;
+        let large_data = prepare_bytes(large_data_size);
+
+        impl<const N: usize> ChunkWithProof<N> {
+            fn replace_first_proof(self) -> Self {
+                let ChunkWithProof { mut proof, chunk } = self;
+
+                let mut merkle_proof: Vec<_> = proof.merkle_proof().to_vec();
+                merkle_proof.swap(0, 1);
+                proof.inject_merkle_proof(merkle_proof);
+
+                ChunkWithProof { proof, chunk }
+            }
+        }
+
+        let chunk_with_proof = ChunkWithProof::<CHUNK_SIZE>::new(large_data.as_slice(), 0).unwrap();
+        assert!(chunk_with_proof.is_valid());
+
+        let chunk_with_incorrect_proof = chunk_with_proof.replace_first_proof();
+        assert!(!chunk_with_incorrect_proof.is_valid());
+    }
+
+    #[test]
+    fn validates_chunk_with_proof_after_deserialization() {
+        let large_data_size = (CHUNK_SIZE * 2) as usize;
+        let large_data = prepare_bytes(large_data_size);
+        let chunk_with_proof = ChunkWithProof::<CHUNK_SIZE>::new(large_data.as_slice(), 0).unwrap();
+
+        let json = serde_json::to_string(&chunk_with_proof).unwrap();
+        serde_json::from_str::<ChunkWithProof<CHUNK_SIZE>>(&json)
+            .expect("should deserialize correctly");
```

Comment:
```suggestion
        assert_eq!(
          chunk_with_proof,
          serde_json::from_str::<ChunkWithProof<CHUNK_SIZE>>(&json)
              .expect("should deserialize correctly"));
```

You could make a proptest...

Reply:
Applied

Reply:
I will turn most or all UT into proptests a bit later, after I'm sure that the UTs respects the expected behavior of the functions.